draft-irtf-cfrg-hybrid-kems.md

```diff
@@ -270,262 +296,371 @@ of testing):
+    return DeriveKeyPair(seed)
+
+def DeriveKeyPair(seed):
+    seed_full = ExpandHash(seed)
+    (seed_T, seed_PQ) = split(Group_T.Nseed, KEM_PQ.Nseed, seed)
+
+    dk_T = Group_T.RandomScalar(seed_T))
+    ek_T = Group_T.Exp(Group_T.g, dk_T)
+    (ek_PQ, dk_PQ) = KEM_PQ.DeriveKeyPair(seed_PQ)
 
-Note that pq_CT and pq_EK are NOT included in the KDF. This is only possible
-because the component KEMs adhere to the following requirements. The QSF
-combiner MUST NOT be used in concrete KEM instances that do not satisfy these
-requirements.
+    ek_H = concat(ek_T, ek_PQ)
+    dk_H = concat(dk_T, dk_PQ)
+    return (ek_H, dk_H)
 
-1. Nominal Diffie-Hellman Group with strong Diffie-Hellman security
+def Encaps(ek):
+    (ek_T, ek_PQ) = split(Group_T.Nek, KEM_PQ.Nek, ek)
 
-A cryptographic group modelable as a nominal group where the strong
-Diffie-Hellman assumption holds {XWING}. Specically regarding a nominal
-group, this means that especially the QSF construction's security is
-based on a computational-Diffie-Hellman-like problem, but no assumption is
-made about the format of the generated group element - no assumption is made
-that the shared group element is indistinguishable from random bytes.
+    sk_E = Group_T.RandomScalar(random(GroupT.nseed))
+    ct_T = Group_T.Exp(GroupT.g, sk_E)
+    ss_T = Group_T.ElementToSharedSecret(Group_T.Exp(ek_T, sk_E))
+    (ss_PQ, ct_PQ) = KEM_PQ.Encap(ek_PQ)
 
-2. Post-quantum IND-CCA KEM with ciphertext second preimage resistance
+    ss_H = CombinerHash(concat(ss_PQ, ss_T, ct_T, ek_T, Label))
+    ct_H = concat(ct_T, ct_PQ)
+    return (ss_H, ct_H)
 
-The QSF relies the post-quantum KEM component having IND-CCA security against
-a post-quantum attacker, and ciphertext second preimage resistance (C2SPI,
-also known as chosen ciphertext resistance, CCR). C2SPI/CCR is [equivalent to
-LEAK-BIND-K,PK-CT security][CDM23]
+def Decaps(dk, ct):
+    (dk_T, dk_PQ) = split(Group_T.Ndk, KEM_PQ.Ndk, dk)
+    (ct_T, ct_PQ) = split(Group_T.Nct, KEM_PQ.Nct, ct)
 
-3. KDF is a secure (post-quantum) PRF, modelable as a random oracle.
+    ek_T = Group_T.ToEncaps(dk_T)
+    ek_PQ = KEM_PQ.ToEncaps(dk_PQ)
 
-Indistinguishability of the final shared secret from a random key is
-established by modeling the key-derivation function as a random
-oracle {{XWING}}.
+    ss_T = Group_T.ElementToSharedSecret(Group_T.Exp(ct_T, dk_T))
+    ss_PQ = KEM_PQ.Decap(dk_PQ, ct_PQ)
+
+    ss_H = CombinerHash(concat(ss_PQ, ss_T, ct_T, ek_T, Label))
+    return ss_H
+~~~
 
 # Security Considerations
 
+## Security Properties
+
+### INDistinguishability against Chosen-Ciphertext Attacks (IND-CCA)
+
+### Ciphertext Second Preimage Resistance (C2PR)
+
+### Binding Properties (X-BIND-P-Q)
+
+### Survival if One KEM Fails
+
+## Security of the Combiners
+
+### Everything {#everything-sec}
+
+### OnlyTraditional {#only-traditional-sec}
+
+### OnlySharedSecrets {#only-shared-secrets-sec}
+
+# Security Considerations (Original)
+
```

Comment:
Removed for now but tracked in https://github.com/cfrg/draft-irtf-cfrg-hybrid-kems/issues/24

```suggestion
```